src/ride/ride.c

```diff
@@ -6171,9 +6170,47 @@ void ride_update_max_vehicles(int rideIndex)
 	}
 }
 
-static void sub_6DE52C(rct_ride *ride)
+/*
+ * rtc2: 0x006DE52C
+ */
+static void ride_reset_vehicle_colours(rct_ride *ride)
 {
-	RCT2_CALLPROC_X(0x006DE52C, 0, 0, 0, 0, (int)ride, 0, 0);
+	uint8* colourCount = GET_RIDE_ENTRY(ride->subtype)->default_colours_ptr;
+	rct_vehicle_colour_extended* colour = colourCount + 1;
+
+	//Different colour per train
+	if (*colourCount == 255) {
+		ride->colour_scheme_type = RIDE_COLOUR_SCHEME_DIFFERENT_PER_TRAIN;
+		for (int index = 0; index < 32; index++) {
+			ride->vehicle_colours[index] = colour[index].vehicle_colour;
+			ride->vehicle_colours_extended[index] = colour[index].restraint_colour;
+		}
+		return;
+	}
+
+	//Same colour per train, different from all existing rides of this type if possible
+	ride->colour_scheme_type = RIDE_COLOUR_SCHEME_ALL_SAME;
+	int remaining = 0xC8;		//Not sure why 200
+	rct_ride* ride_list = RCT2_ADDRESS(RCT2_ADDRESS_RIDE_LIST, rct_ride);
+	uint16 colourIndex;
+	rct_ride* rideTemp;
+
+getRandomColorScheme:
+	if (--remaining != 0) {
+		colourIndex = ((uint16)*colourCount * (uint16)(uint8)scenario_rand()) >> 8;	//High bit of a 8bit x 8bit mul operation
+		rct_vehicle_colour v_colour = colour[colourIndex].vehicle_colour;
+		for (int index = 0; index < 255; index++) {	//Check to see if any rides of this subtype already have this color scheme
+			rideTemp = &ride_list[index];
+			if (rideTemp->type != 255 && rideTemp != ride && ride->subtype == rideTemp->subtype && v_colour.vehicle_colour == rideTemp->vehicle_colours[0].vehicle_colour) {
```

Comment:
There is far too much in this one line. I'd prefer it if you broke it up first checking to make sure the ride is valid, then checking its not the current ride then...